functions/go/set-namespace/transformer/namespace.go

```diff
@@ -98,45 +71,152 @@ func (p *NamespaceTransformer) Config(o *fn.KubeObject) error {
+// - Restrict mode: always update namespace from the given `fieldpaths`
+// - NsObjectSelector mode: update the namespace which matches the Namespace object found in resourceList.items.
+// - CustomSelector mode: update the namespace which matches the `namespaceSelector` in resourceList.functionConfig.
+func (p *NamespaceTransformer) updateNamespace(o *fn.KubeObject, fieldPaths ...string) {
+	if p.isSelectorMode() {
+		ns := o.GetStringOrDie(fieldPaths...)
+		if ns != p.MatchingNamespace {
+			return
+		}
+	}
+	if err := o.Set(p.NewNamespace, fieldPaths...); err != nil {
+		p.Errors = append(p.Errors, err.Error())
+	}
+}
+
+// isSelectorMode decides whether the namespace fields should match the given namespace
+// before updating.
+func (p *NamespaceTransformer) isSelectorMode() bool {
+	switch p.Mode {
+	case CustomSelector:
+		return true
+	case NsObjectSelector:
+		return true
+	default: //  Restrict Mode
+		return false
+	}
+}
+
+// validate guarantees the input resourceList.items satisfy the UpdateMode requirements.
+func (p *NamespaceTransformer) validate(objects []*fn.KubeObject) bool {
+	nsCount, fromResources := p.listCurrentNamespaces(objects)
+	// Check "CustomSelector" Mode
+	if p.MatchingNamespace != "" {
+		p.Mode = CustomSelector
+		if len(nsCount) == 0 {
+			return true
+		}
+		p.Errors = append(p.Errors,
+			"found Namespace objects from the input resources, "+
+				"you cannot use `namespaceSelector` in FunctionConfig together with Namespace objects")
+		return false
+	}
+	// Check "NsObjectSelector" Mode
+	if len(nsCount) > 1 {
+		msg := fmt.Sprintf("cannot accept more than one Namespace objects from the input resources, found %v",
+			nsCount)
+		p.Errors = append(p.Errors, msg)
+		return false
+	}
+	if len(nsCount) == 1 {
+		matchingNs := reflect.ValueOf(nsCount).MapKeys()[0].String()
+		if nsCount[matchingNs] > 1 {
+			msg := fmt.Sprintf("found more than one Namespace objects of the same name %v", matchingNs)
+			p.Errors = append(p.Errors, msg)
+			return false
+		}
+		p.Mode = NsObjectSelector
+		p.MatchingNamespace = matchingNs
+		return true
+	}
+	// Check "Restrict" Mode
+	if len(fromResources) > 1 {
+		msg := fmt.Sprintf("all input namespace-scoped resources should be under the same namespace "+
+			"but found different namespaces: %v ", reflect.ValueOf(fromResources).MapKeys())
+		p.Errors = append(p.Errors, msg)
+		return false
+	}
+	p.Mode = Restrict
+	p.MatchingNamespace = reflect.ValueOf(fromResources).MapKeys()[0].String()
+	return true
+}
+
+// listCurrentNamespaces iterates the input resourcelist.items and list all namespaces found.
+func (p *NamespaceTransformer) listCurrentNamespaces(objects []*fn.KubeObject) (map[string]int, map[string]bool) {
+	nsCount := map[string]int{}
+	fromResources := map[string]bool{}
+	for _, o := range objects {
+		// Skip local resource which `kpt live apply` skips.
+		if o.IsLocalConfig() {
+			continue
+		}
+		switch {
+		case o.IsGVK("v1", "Namespace"):
```

Comment:
I think it would be great to combine the logic with the transformation logic.

A good starting point might be something like:

`func VisitNamespaces(objects []*fn.KubeObject, visitor func (ns string) string)`

(I can see us replacing `ns string` with something like `v Namespace` and supporting `v.Set()`, `v.Node()`, `v.FieldPath()` etc, but I don't think we need to do that yet)